[GTK4] Gamepad tests are timing out when run with run-webkit-tests

carlosgcampos opened the pull request:
#### ff9142f913ffc295a69db52ccf83c29a29cf0e64
<pre>
[GTK4] Gamepad tests are timing out when run with run-webkit-tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=279016">https://bugs.webkit.org/show_bug.cgi?id=279016</a>

Reviewed by Adrian Perez de Castro.

Tests timeout because we fail to get the toplevel active window under
Xvfb that doesn&apos;t support toplevel focus. We can always assume that the
web view toplevel window is focused in that case since it&apos;s the only
one.

Canonical link: <a href="https://commits.webkit.org/297162@main">https://commits.webkit.org/297162@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c51a75217eea06fea722f2427b9aadb002af82d9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30248 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116615 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60856 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38837 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84098 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113537 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24696 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99581 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64539 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24059 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17721 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60410 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94082 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17780 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119405 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37629 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27945 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93063 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38003 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95852 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92885 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37897 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15641 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33604 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17861 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37524 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42995 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37187 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40526 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38895 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c51a75217eea06fea722f2427b9aadb002af82d9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47710

csaavedra commented on Source/WebKit/UIProcess/Gamepad/gtk/UIGamepadProviderGtk.cpp:
```diff
@@ -68,22 +68,34 @@ static WebPageProxy* getWebPageProxy(GtkWidget* widget)
 
+static bool windowHasTopLevelFocus(GtkWindow* window)
+{
+#if ENABLE(DEVELOPER_MODE)
+    // Xvfb doesn't support toplevel focus, so we consider toplevel window to be always active since it's the only one.
+    if (Display::singleton().isX11()) {
+        static const char* underXvfb = g_getenv("UNDER_XVFB");
+        if (!g_strcmp0(underXvfb, "yes"))
+            return true;
+    }
+#endif
```
Can't this go to a single place in some way? I see this comment in at least 4 or 5 places with the same checks.

carlosgcampos replied:
Yes, but in a follow up patch.

aperezdc replied:
Yes, please 🙏🏼

carlosgcampos committed:
[GTK4] Gamepad tests are timing out when run with run-webkit-tests
https://bugs.webkit.org/show_bug.cgi?id=279016

Reviewed by Adrian Perez de Castro.

Tests timeout because we fail to get the toplevel active window under
Xvfb that doesn't support toplevel focus. We can always assume that the
web view toplevel window is focused in that case since it's the only
one.

Canonical link: https://commits.webkit.org/297162@main
